util/saconfig

```diff
@@ -61,74 +62,102 @@ def calibrate(dev, channels):
         logging.info("Calibrating channel %d", channel)
         # set calibration options
         dev.SetProperty_i32(smaract.SA_CTLDLL.SA_CTL_PKEY_CALIBRATION_OPTIONS,
-                        channel, calibration_options)
+                            channel, calibration_options)
 
         dev.Calibrate(channel)
 
 
-def autozero(dev, channels):
+DISTANCE_CODED_MARK_POSITIONER = 304
+
+
+def set_positioner_type(dev, channels, ptype):
+    """
+    Set the type of positioner that is connected to the device channels. The positioner type implicitly gives the
+    controller information about how to calculate positions, handle the referencing, configure the control-loop,
+    etc. See sections 2.5 and 4.4.6 in programmer manual.
+    """
     if not channels:
         return
+    for channel in channels:
+        logging.info(f"Setting positioner type for channel {channel} to {ptype}.")
+        dev.SetProperty_i32(smaract.SA_CTLDLL.SA_CTL_PKEY_POSITIONER_TYPE, channel,
+                            ptype)
 
-    # Set auto-zero mode
+
+def set_reference_autozero(dev, channels):
+    if not channels:
+        return
+    # Set auto-zero referencing mode
     for channel in channels:
-        logging.info("Setting auto-zero mode for channel %d", channel)
+        logging.info(f"Setting auto-zero mode for channel {channel}")
         dev.SetProperty_i32(smaract.SA_CTLDLL.SA_CTL_PKEY_REFERENCING_OPTIONS,
                             channel, smaract.SA_CTLDLL.SA_CTL_REF_OPT_BIT_AUTO_ZERO)  # auto zero
 
-    # Run the referencing
-    logging.info("Starting reference...")
-    dev.reference().result()
-    time.sleep(0.1)
-    logging.info("Complete")
+    reference(dev, channels)
+
 
+def set_reference_normal(dev, channels):
+    if not channels:
+        return
     # Set normal referencing mode
-    for channel in  channels:
-        logging.info("Setting the reference mode to normal for channel %d", channel)
+    for channel in channels:
+        logging.info(f"Setting the reference mode to normal for channel {channel}")
         dev.SetProperty_i32(smaract.SA_CTLDLL.SA_CTL_PKEY_REFERENCING_OPTIONS,
                             channel, smaract.SA_CTLDLL.SA_CTL_REF_OPT_BIT_NORMAL)  # normal
 
 
+def reference(dev, channels):
+    # Run the referencing
+    logging.info("Starting reference...")
+    dev.reference(axes=set([str(ch) for ch in channels])).result()
+    time.sleep(0.1)
+    logging.info("Referencing complete.")
+    # Always set reference to normal after referencing
+    set_reference_normal(dev, channels)
```

Comment:
Is that what you had in mind? @pieleric  `reference()` is called from `set_reference_autozero()` which means the reference mode will return back to normal.

Reply:
Not exactly. Right now it's confusing because set_reference_autozero() does the referencing, while set_reference_normal() just change the flag. We end up with set_reference_autozero() calling indirectly set_reference_normal().
I'd suggest to add a flag to `reference():` `autozero=False`. For each channel, it would set/unset the autozero bit. Note that it should just change that bit, not change the whole value. You need to use bitwise and/or! So it could rely on a helper subfunction set_reference_autozero(autozero) to do it. Something like this:
```python
def reference(dev, channels, autozero=False):
    for channel in channels:
         set_reference_autozero(autozero)
    # Run the referencing
    logging.info("Starting reference...")
    try:
        dev.reference(axes=set(str(ch) for ch in channels)).result()
        time.sleep(0.1)
        logging.info("Referencing complete.")
    finally:
        # Always set reference to normal after referencing
    for channel in channels:
         set_reference_autozero(False)


def set_reference_autozero(dev, channel, autozero):
    logging.info(f"Setting the reference autozero mode to {autozero} for channel {channel}")
    ref_opt = dev.GetProperty_i32(smaract.SA_CTLDLL.SA_CTL_PKEY_REFERENCING_OPTIONS, channel)
    if autozero:
        ref_opt |= smaract.SA_CTLDLL.SA_CTL_REF_OPT_BIT_AUTO_ZERO
    else:
        ref_opt &= ~smaract.SA_CTLDLL.SA_CTL_REF_OPT_BIT_AUTO_ZERO

```